instrumentation/micrometer/micrometer-1.5/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java

```diff
@@ -0,0 +1,167 @@
+
+  // TODO: use bound instruments when they're available
+  private final LongHistogram otelHistogram;
+  private final Attributes attributes;
+  private final Measurements measurements;
+
+  volatile boolean removed = false;
+
+  OpenTelemetryTimer(
+      Id id,
+      Clock clock,
+      DistributionStatisticConfig distributionStatisticConfig,
+      PauseDetector pauseDetector,
+      Meter otelMeter) {
+    super(id, clock, distributionStatisticConfig, pauseDetector, TimeUnit.MILLISECONDS, false);
+
+    this.otelHistogram =
+        otelMeter
+            .histogramBuilder(id.getName())
+            .setDescription(description(id))
+            .setUnit("milliseconds")
+            .ofLongs()
+            .build();
+    this.attributes = toAttributes(id.getTags());
+
+    if (isUsingMicrometerHistograms()) {
+      measurements = new MicrometerHistogramMeasurements(clock, distributionStatisticConfig);
+    } else {
+      measurements = NoopMeasurements.INSTANCE;
+    }
+  }
+
+  boolean isUsingMicrometerHistograms() {
+    return histogram != NoopHistogram.INSTANCE;
+  }
+
+  @Override
+  protected void recordNonNegative(long amount, TimeUnit unit) {
+    if (amount >= 0 && !removed) {
+      otelHistogram.record(unit.toMillis(amount), attributes);
+      measurements.record(amount, unit);
+    }
+  }
+
+  @Override
+  public long count() {
+    return measurements.count();
+  }
+
+  @Override
+  public double totalTime(TimeUnit unit) {
+    return measurements.totalTime(unit);
+  }
+
+  @Override
+  public double max(TimeUnit unit) {
+    return measurements.max(unit);
+  }
+
+  @Override
+  public Iterable<Measurement> measure() {
+    // OpenTelemetry metrics bridge does not support reading measurements
+    return Collections.emptyList();
+  }
+
+  @Override
+  public void onRemove() {
+    removed = true;
+  }
+
+  private interface Measurements {
+    void record(long amount, TimeUnit unit);
+
+    long count();
+
+    double totalTime(TimeUnit unit);
+
+    double max(TimeUnit unit);
+  }
+
+  // if micrometer histograms are not being used then there's no need to keep any local state
+  // OpenTelemetry metrics bridge does not support reading measurements
+  enum NoopMeasurements implements Measurements {
+    INSTANCE;
+
+    @Override
+    public void record(long amount, TimeUnit unit) {}
+
+    @Override
+    public long count() {
+      return 0;
+    }
+
+    @Override
+    public double totalTime(TimeUnit unit) {
+      return Double.NaN;
+    }
+
+    @Override
+    public double max(TimeUnit unit) {
+      return Double.NaN;
+    }
+  }
+
+  // calculate count, totalTime and max value for the use of micrometer histograms
+  // kinda similar to how DropwizardTimer does that
+  private static final class MicrometerHistogramMeasurements implements Measurements {
+
+    private final AtomicLong count = new AtomicLong(0);
+    private final AtomicLong totalTime = new AtomicLong(0);
```

Comment:
Do we need to be worried about performance problems with high contention here? Should we use `LongAdder` instead?

The usage of `TmeWindowMax` suggests that performance is a consideration.

Reply:
Micrometer itself is a bit inconsistent here: `DropwizardTimer` uses `AtomicLong` for total time, but `StepTimer` uses `LongAdder`s. Perhaps it doesn't really matter that much for the dropwizard integration? Anyway, using adders is probably better here, so I'll do just that.